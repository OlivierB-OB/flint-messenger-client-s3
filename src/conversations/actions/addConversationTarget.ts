import { IAddConversationTargetAction, ADD_CONVERSATION_TARGET } from '../types';

export function addConversationTarget(conversationId: string, target: string): IAddConversationTargetAction {
  return {
    type: ADD_CONVERSATION_TARGET,
    conversationId,
    target,
  };
}
