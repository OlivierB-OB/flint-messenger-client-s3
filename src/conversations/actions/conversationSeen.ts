import { IConversationSeenAction, CONVERSATION_SEEN } from '../types';

export function conversationSeen(id: string, seenDate: string): IConversationSeenAction {
  return {
    type: CONVERSATION_SEEN,
    id,
    seenDate,
  };
}
