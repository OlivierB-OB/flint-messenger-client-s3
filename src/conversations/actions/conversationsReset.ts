import { IConversationsResetAction, CONVERSATIONS_RESET } from '../types';

export function conversationsReset(): IConversationsResetAction {
  return { type: CONVERSATIONS_RESET };
}
