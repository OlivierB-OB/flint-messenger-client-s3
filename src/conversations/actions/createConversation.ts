import { ICreateConversationAction, CREATE_CONVERSATION } from '../types';

export function createConversation(conversationId: string, targets: string[]): ICreateConversationAction {
  return {
    type: CREATE_CONVERSATION,
    conversationId,
    targets,
  };
}
