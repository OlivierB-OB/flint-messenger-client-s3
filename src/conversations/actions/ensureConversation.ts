import { IEnsureConversationAction, ENSURE_CONVERSATION } from '../types';

export function ensureConversation(
  conversationId: string,
  targets: string[],
  createdAt: string,
): IEnsureConversationAction {
  return {
    type: ENSURE_CONVERSATION,
    conversationId,
    targets,
    createdAt,
  };
}
