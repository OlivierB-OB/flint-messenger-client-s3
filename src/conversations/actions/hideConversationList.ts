import { IToggleConversationListAction, TOGGLE_CONVERSATION_LIST } from '../types';

export function hideConversationList(): IToggleConversationListAction {
  return {
    type: TOGGLE_CONVERSATION_LIST,
    value: false,
  };
}
