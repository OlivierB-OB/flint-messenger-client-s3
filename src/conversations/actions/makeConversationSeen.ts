import axios from 'axios';
import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { IConversationSeenAction, CONVERSATION_SEEN } from '../types';
import { updateIdentity } from '../../identity/actions/updateIdentity';

function conversationSeen(id: string, seenDate: string): IConversationSeenAction {
  return {
    type: CONVERSATION_SEEN,
    id,
    seenDate,
  };
}

export function makeConversationSeen(id: string) {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    const seenDate = new Date().toISOString();
    
    try {
      const response = await axios.patch(`${process.env.REACT_APP_BACKEND}/profile/conversation-seen`, {
        conversationId: id,
        seenDate,
      }, { withCredentials: true });
      batch(() => {
        dispatch(updateIdentity(response.data));
        dispatch(conversationSeen(id, seenDate));
      })
    } catch (error) {
      // IGNORE...
    }
  };
}
