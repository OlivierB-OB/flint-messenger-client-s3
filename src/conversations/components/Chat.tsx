import Box from '@material-ui/core/Box';
import Container from '@material-ui/core/Container';
import LinearProgress from '@material-ui/core/LinearProgress';
import React from 'react';
import { connect } from 'react-redux';
import { match as Match } from 'react-router-dom';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { ChatMessages } from './ChatMessages';
import { ChatInput } from './ChatInput';
import { IAppState } from '../../appReducer';
import { IConversation, IConversationsStatus } from '../types';
import { Alert } from '../../layout/components/Alert';
import { makeConversationSeen } from '../actions/makeConversationSeen';

export interface IChatDisplayProps {
  status: IConversationsStatus;
  conversationId?: string;
  conversation?: IConversation;
  conversationSeen: (id: string) => void;
}

export function ChatDisplay({ status, conversationId, conversation, conversationSeen }: IChatDisplayProps) {
  if (!conversationId || !conversation) return null;
  const progress = status === 'sending' ? <LinearProgress /> : null;
  return (
    <Container style={{ height: '100%', padding: '1rem', boxSizing: 'border-box' }}>
      <Box style={{ height: '5%' }}>
        <Alert status={status} />
        {progress}
      </Box>
      <Box style={{ height: '70%', overflow: 'auto' }}>
        <ChatMessages
          conversationId={conversationId}
          messages={conversation.messages}
          conversationSeen={conversationSeen}
        />
      </Box>
      <Box style={{ height: '20%' }}>
        <ChatInput conversationId={conversationId} />
      </Box>
    </Container>
  );
}

export interface IChatProps {
  match: Match<{ conversationId: string }>;
}

const mapStateToProps = ({ conversations }: IAppState, { match }: IChatProps) => ({
  status: conversations.status,
  conversationId: match.params?.conversationId,
  conversation: conversations.conversations.find(({ _id }) => {
    return _id === match.params?.conversationId;
  }),
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  conversationSeen: (id: string) => void dispatch(makeConversationSeen(id)),
});

export const Chat = connect(mapStateToProps, mapDispatchToProps)(ChatDisplay);
