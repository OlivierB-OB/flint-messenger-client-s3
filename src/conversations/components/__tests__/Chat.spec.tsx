import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { Chat } from '../Chat';
import { makeConversationSeen } from '../../actions/makeConversationSeen';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Chat', () => {
  let Connected: any = Chat;
  let spyConversationSeen: jest.Mock;
  let restoreConversationSeen: () => void;

  beforeEach(() => {
    [spyConversationSeen, restoreConversationSeen] = makeConversationSeen.mock(jest.fn());
  });

  afterEach(() => {
    restoreConversationSeen();
  });

  it('should provide status, conversationId, conversation and callbacks to ChatDisplay', async () => {
    const store = mockStore({
      conversations: {
        status: 'conversation status',
        conversations: [{ _id: '123', fake: 'the conversation' }],
      },
    });
    const match: any = {
      params: {
        conversationId: '123',
      },
    };
    const component = shallow(<Connected store={store} match={match} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('conversationSeen') as any)(component.prop('conversationId'));
    expect(spyConversationSeen).toHaveBeenCalledWith('123');
  });

  it('should not crash if the match is not clean', async () => {
    const store = mockStore({
      conversations: {
        status: 'conversation status',
        conversations: [{ _id: '123', fake: 'the conversation' }],
      },
    });
    const match: any = {};
    const component = shallow(<Connected store={store} match={match} />).dive();
    expect(component).toMatchSnapshot();
  });

  it('should not crash if the conversation is not found', async () => {
    const store = mockStore({
      conversations: {
        status: 'conversation status',
        conversations: [{ _id: '123', fake: 'the conversation' }],
      },
    });
    const match: any = {
      params: {
        conversationId: '222',
      },
    };
    const component = shallow(<Connected store={store} match={match} />).dive();
    expect(component).toMatchSnapshot();
  });
});
