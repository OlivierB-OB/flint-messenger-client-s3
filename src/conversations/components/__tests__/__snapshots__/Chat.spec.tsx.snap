// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should not crash if the conversation is not found 1`] = `
<ChatDisplay
  conversationId="222"
  conversationSeen={[Function]}
  match={
    Object {
      "params": Object {
        "conversationId": "222",
      },
    }
  }
  status="conversation status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`Chat should not crash if the match is not clean 1`] = `
<ChatDisplay
  conversationSeen={[Function]}
  match={Object {}}
  status="conversation status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`Chat should provide status, conversationId, conversation and callbacks to ChatDisplay 1`] = `
<ChatDisplay
  conversation={
    Object {
      "_id": "123",
      "fake": "the conversation",
    }
  }
  conversationId="123"
  conversationSeen={[Function]}
  match={
    Object {
      "params": Object {
        "conversationId": "123",
      },
    }
  }
  status="conversation status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
